src/policy/v3_policy.h

```diff
@@ -0,0 +1,84 @@
+ * V3_CHILD_MAX_VSIZE.
+ *
+ *
+ * @param[in]   mempool_ancestors       The in-mempool ancestors of ptx, including any that are only
+ *                                      direct ancestors of its in-package ancestors.
+ * @param[in]   num_in_pkg_ancestors    The number of ancestors not accounted for in ancestors, i.e.
+ *                                      in-package parents that have not been submitted to the mempool yet.
+ * @param[in]   direct_conflicts        In-mempool transactions this tx conflicts with. These conflicts
+ *                                      are used to more accurately calculate the resulting descendant
+ *                                      count of in-mempool ancestors.  While V3_ANCESTOR_LIMIT is 2, it
+ *                                      is unnecessary to include the conflicts of in-package ancestors
+ *                                      because the presence of both in-mempool and in-package ancestors
+ *                                      would already be a violation of V3_ANCESTOR_LIMIT.
+ * @param[in]   vsize                   The sigop-adjusted virtual size of ptx.
+ * @returns an error string if any v3 rule was violated, otherwise std::nullopt.
+ */
+std::optional<std::string> ApplyV3Rules(const CTransactionRef& ptx,
+                                        const CTxMemPool::setEntries& mempool_ancestors,
+                                        unsigned int num_in_pkg_ancestors,
+                                        const std::set<Txid>& direct_conflicts,
+                                        int64_t vsize);
+
+/** Must be called for every package containing any v3 transaction. Should not be called for non-v3 packages.
+ *
+ * Check the following rules for transactions within the package:
+ * 1. A v3 tx must only have v3 unconfirmed ancestors.
+ * 2. A non-v3 tx must only have non-v3 unconfirmed ancestors.
+ * 3. A v3's ancestor set, including itself, must be within V3_ANCESTOR_LIMIT.
+ * 4. A v3's descendant set, including itself, must be within V3_DESCENDANT_LIMIT.
+ * 5. If a v3 tx has any unconfirmed ancestors, its vsize must be within V3_CHILD_MAX_VSIZE.
+ *
+ * Important: this function is necessary but insufficient to enforce these rules. ApplyV3Rules must
+ * be called for each individual transaction, after in-mempool ancestors, virtual sizes, and
+ * in-package ancestors have been calculated. This function serves as a way to quit early on
+ * packages in which those calculations may be expensive.
+ *
+ * @returns If all checks pass, a map from each v3 transaction, by txid, to the txids of its
+ * in-package ancestor set. Every ancestor set includes the tx itself. If this passed, we know that
+ * each connected component does not violate V3 inheritance or topology constraints within the
+ * package itself. If any checks fail, an error string detailing what failed.
+ * */
+util::Result<std::map<Txid, std::set<Txid>>> PackageV3Checks(const Package& package);
```

Comment:
Tbh since the limit is 1 ancestor, we could just make this a `std::map<Txid, Txid>` and save some space.